_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,332 @@
+
+#### Literal Suffixes
+The very next step is the ability to provide custom literals of user-defined value types.
+The ease of using literal values in expressions is a big part of what makes a type feel native.
+Solidity already provides fractional literals and our intention is to make them usable with custom
+fixed-point number implementations.
+To achieve that we are going to allow defining suffix functions that return UDVT by converting
+literals that are already available in the language:
+
+```solidity
+type UFixed16x2 is uint16;
+
+function uf(int16 mantissa, uint8 exponent) pure suffix returns (UFixed16x2) { ... }
+```
+```solidity
+UFixed16x2 circumference = 2 uf * 3.14 uf * radius;
+```
+
+While the application of a suffix will involve a function call, eventually, with the help of
+compile-time evaluation, literal suffixes are going to become as efficient as native literals.
+
+#### More Operators
+Several operators have been excluded from the initial implementation for the sake of simplicity:
+1. `&&`, `||` and `!` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    In case of `&&` and `||`  an additional complication is their short-circuiting behavior.
+1. `<<`, `>>` and `**` have non-uniform arguments.
+    For built-in types, the right-hand side is always an unsigned type.
+1. `=` is one of the few operators that already worked with UDVTs.
+    The new feature is not meant to allow redefining existing operators so providing a custom
+    definition for it will not be possible.
+1. Other assignment operators, as `+=`, for example, modify their argument and return a reference.
+    There are currently no references to value types in the language which makes implementing them
+    directly impossible.
+    It may also be desirable to have the compiler provide them automatically whenever the
+    corresponding bitwise or arithmetic operator is defined on the type.
+1. `++` and `--` modify their argument and also have a separate prefix and postfix variants with
+    different semantics.
+
+We are considering enabling them in one of the future releases, provided that the design issues can
+be resolved satisfactorily.
+There are use cases where operators like `++`, `--`, or `**` would be a natural fit, so they are
+very likely to become available sooner or later.
+
+#### Operators on Structs
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
+The main thing holding us back are the planned changes to reference types, in particular the fact
+that data location (`storage`, `memory`, etc.) is going to become a part of the type itself.
+To avoid having to introduce breaking changes later, we decided to defer support for structs until
+these changes are behind us.
```

Comment:
```suggestion
While UDVTs were the primary target, not all types can reasonably fit into a single slot.
The way to solve this problem is to allow operators to take and return structs.
The main thing holding us back are the planned changes to reference types, in particular the fact
that data location (`storage`, `memory`, etc.) is going to become a part of the type itself.
To avoid having to introduce breaking changes later, we decided to defer support for structs until
these changes are behind us.
```